vendor/lunasvg/3rdparty/software/sw_ft_stroker.c

```diff
@@ -0,0 +1,1937 @@
+      SW_FT_Vector_From_Polar( &a3, radius,
+                            angle_start + i * angle_diff / arcs );
+      a2.x = SW_FT_MulFix(  a3.y, coef );
+      a2.y = SW_FT_MulFix( -a3.x, coef );
+
+      a3.x += center->x;
+      a3.y += center->y;
+      a2.x += a3.x;
+      a2.y += a3.y;
+
+      /* add cubic arc */
+      error = ft_stroke_border_cubicto( border, &a1, &a2, &a3 );
+      if ( error )
+        break;
+
+      /* a0 = a3; */
+      a1.x = a3.x - a2.x + a3.x;
+      a1.y = a3.y - a2.y + a3.y;
+    }
+
+    return error;
+}
+
+static SW_FT_Error ft_stroke_border_moveto(SW_FT_StrokeBorder border,
+                                           SW_FT_Vector*      to)
+{
+    /* close current open path if any ? */
+    if (border->start >= 0) ft_stroke_border_close(border, FALSE);
+
+    border->start = border->num_points;
+    border->movable = FALSE;
+
+    return ft_stroke_border_lineto(border, to, FALSE);
+}
+
+static void ft_stroke_border_init(SW_FT_StrokeBorder border)
+{
+    border->points = NULL;
+    border->tags = NULL;
+
+    border->num_points = 0;
+    border->max_points = 0;
+    border->start = -1;
+    border->valid = FALSE;
+}
+
+static void ft_stroke_border_reset(SW_FT_StrokeBorder border)
+{
+    border->num_points = 0;
+    border->start = -1;
+    border->valid = FALSE;
+}
+
+static void ft_stroke_border_done(SW_FT_StrokeBorder border)
+{
+    free(border->points);
+    free(border->tags);
+
+    border->num_points = 0;
+    border->max_points = 0;
+    border->start = -1;
+    border->valid = FALSE;
+}
+
+static SW_FT_Error ft_stroke_border_get_counts(SW_FT_StrokeBorder border,
+                                               SW_FT_UInt*        anum_points,
+                                               SW_FT_UInt*        anum_contours)
+{
+    SW_FT_Error error = 0;
+    SW_FT_UInt  num_points = 0;
+    SW_FT_UInt  num_contours = 0;
+
+    SW_FT_UInt    count = border->num_points;
+    SW_FT_Vector* point = border->points;
+    SW_FT_Byte*   tags = border->tags;
+    SW_FT_Int     in_contour = 0;
+
+    for (; count > 0; count--, num_points++, point++, tags++) {
+        if (tags[0] & SW_FT_STROKE_TAG_BEGIN) {
+            if (in_contour != 0) goto Fail;
+
+            in_contour = 1;
+        } else if (in_contour == 0)
+            goto Fail;
+
+        if (tags[0] & SW_FT_STROKE_TAG_END) {
+            in_contour = 0;
+            num_contours++;
+        }
+    }
+
+    if (in_contour != 0) goto Fail;
+
+    border->valid = TRUE;
+
+Exit:
+    *anum_points = num_points;
+    *anum_contours = num_contours;
+    return error;
+
+Fail:
+    num_points = 0;
+    num_contours = 0;
+    goto Exit;
+}
+
+static void ft_stroke_border_export(SW_FT_StrokeBorder border,
+                                    SW_FT_Outline*     outline)
+{
+    /* copy point locations */
+    memmove(outline->points + outline->n_points, border->points,
+           border->num_points * sizeof(SW_FT_Vector));
```

Comment:
Let's make a distinction here that we've changed the original vendor library source code, so that we don't lose them during future upgrades.
```suggestion
    /** Changed by MTA */
    /* move point locations */
    memmove(outline->points + outline->n_points, border->points,
           border->num_points * sizeof(SW_FT_Vector));
    /** */
```

Reply:
Added comments for both the memcpy -> memmove change and assertion removal